### PR TITLE
docker: support shared network namespaces

### PR DESCRIPTION
**Pull request by nickethier**

### CODE REVIEW

File: drivers/docker/network.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+const dockerNetSpecLabelKey = "docker_sandbox_container_id"
+
+func (d *Driver) CreateNetwork(allocID string) (*drivers.NetworkIsolationSpec, error) {
+	// Initialize docker API clients
+	client, _, err := d.dockerClients()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to docker daemon: %s", err)
+	}
+
+	repo, _ := parseDockerImage(infraContainerImage)
```
Comment (schmichael) [inline drivers/docker/network.go]: Check the error in case we make infraContainerImage configurable someday.

File: drivers/docker/network.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		return fmt.Errorf("failed to connect to docker daemon: %s", err)
+	}
+
+	return client.RemoveContainer(docker.RemoveContainerOptions{
+		Force: true,
+		ID:    spec.Labels[dockerNetSpecLabelKey],
+	})
+}
+
+func (d *Driver) createSandboxContainerConfig(allocID string) (*docker.CreateContainerOptions, error) {
```
Comment (schmichael) [inline drivers/docker/network.go]: comment

File: drivers/docker/network.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+	return &docker.CreateContainerOptions{
+		Name: fmt.Sprintf("nomad_%s", allocID),
```
Comment (schmichael) [inline drivers/docker/network.go]: Let's not presume this is the last container Nomad runs on behalf of the user in an alloc :sweat_smile:  (Name is up for debate: could just call it `nomad_pause_$ID` or `nomad_init_$ID`)
```suggestion
		Name: fmt.Sprintf("nomad_net_%s", allocID),
```

File: drivers/docker/network.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		Config: &docker.Config{
+			Image: infraContainerImage,
+		},
+		HostConfig: &docker.HostConfig{
+			NetworkMode: "none",
```
Comment (schmichael) [inline drivers/docker/network.go]: Might want specifically comment this line as I had to go check the Docker docs for it to make sense. The docs even start kind of confusing:

> If you want to completely disable the networking stack on a container, ...

!!! That seems bad! But then:

> Within the container, only the loopback device is created.

So now it makes sense: "disabling the network stack" seems to mean "no bridge, no forwarding, just a netns with a loopback" which is exactly what we want?

File: drivers/rawexec/driver.go [hunk 5]
```diff
@@ -95,6 +95,10 @@ var (
 		SendSignals: true,
 		Exec:        true,
 		FSIsolation: drivers.FSIsolationNone,
+		NetIsolationModes: []drivers.NetIsolationMode{
+			drivers.NetIsolationModeHost,
+			drivers.NetIsolationModeGroup,
```
Comment (schmichael) [inline drivers/rawexec/driver.go]: Does this mean rawexec will have netns support? Or will it just allow you to have a rawexec task that runs on the host in a group with other tasks that share a netns?

File: drivers/docker/config.go [hunk 6]
```diff
@@ -310,6 +310,12 @@ var (
 		SendSignals: true,
 		Exec:        true,
 		FSIsolation: drivers.FSIsolationImage,
+		NetIsolationModes: []drivers.NetIsolationMode{
+			drivers.NetIsolationModeHost,
+			drivers.NetIsolationModeGroup,
+			drivers.NetIsolationModeTask,
```
[the remark this answers is not in the document]
Comment (nickethier) [reply drivers/docker/config.go]: A `none` isolation mode is different than a `none` docker network mode. For the isolation mode, none signals that there is no network to isolate. In practice we don't use this isolation mode on any drivers today but is there if a driver's runtime is remote from the client.

The in the `none` docker network mode there is still network isolation. A namespace is created but no interfaces (except for loopback) are added thus it joins no networks.

File: drivers/docker/network.go (hunk 1 above)
Comment (nickethier) [reply drivers/docker/network.go]: This actually doesn't return an error. It splits the image name (assigned to `repo` here) and tag.

File: drivers/docker/network.go (hunk 4 above)
Comment (nickethier) [reply drivers/docker/network.go]: Yup you got it! Added a comment

File: drivers/rawexec/driver.go (hunk 5 above)
Comment (nickethier) [reply drivers/rawexec/driver.go]: Yup you can run raw_exec tasks in an alloc network.

File: drivers/docker/driver.go [hunk 7]
```diff
@@ -860,9 +860,14 @@ func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *T
...
 	hostConfig.ReadonlyRootfs = driverConfig.ReadonlyRootfs
 
+	// set the docker network mode
 	hostConfig.NetworkMode = driverConfig.NetworkMode
+
+	// if the driver config does not specify a network mode then try to use the
+	// shared alloc network
 	if hostConfig.NetworkMode == "" {
-		if task.NetworkIsolation.Path != "" {
+		if task.NetworkIsolation != nil && task.NetworkIsolation.Path != "" {
```
Comment (schmichael) [inline drivers/docker/driver.go]: I like helpers for these cases as it's a clearer signal to developers that a nil check is necessary. Something like:

```go
if task.NetworkIsolation.HasPath()
```

`Deployment` has a bunch of helpers like this. I don't think it's always necessary, so not a blocker (can also be done later).

**Pull request merged**